### PR TITLE
NCD-343: Read configuration, merge with defaults, show in GUI

### PR DESCRIPTION
**Pull request by cybic**
* Read configuration from firmware, using nrfutil-device commands in shared.
* Wrap configuration and merge with boards defaults from the board definition file
* Set current hardware configuration in state to update GUI

### CODE REVIEW

File: src/SidePanel.tsx [hunk 1]
```diff
@@ -51,6 +51,29 @@ export default () => {
                 >
                     Write config
                 </Button>
+                <Button
+                    disabled={!device}
+                    variant="secondary"
+                    className="w-100"
+                    onClick={async () => {
+                        console.dir(device);
```
Comment (boundlesscalm) [inline src/SidePanel.tsx:59]: Leftover from debugging?

File: src/app/DeviceSelector.tsx [hunk 2]
```diff
@@ -59,9 +62,10 @@ const deviceListing: DeviceTraits = {
  * Note that the callbacks releaseCurrentDevice and onDeviceIsReady
  * are only invoked, if a deviceSetup is defined.
  */
-const onDeviceSelected = (dispatch: AppDispatch) => (device: Device) => {
+const onDeviceSelected = (dispatch: AppDispatch) => async (device: Device) => {
     logger.info(`Selected device with s/n ${device.serialNumber}`);
-    getBoardControllerVersion(dispatch, device); // FIXME: Remove this when onDeviceIsReady() is called
+    await getBoardControllerVersion(dispatch, device); // FIXME: Remove this when onDeviceIsReady() is called
+    await getCurrentBoardControllerConfig(dispatch, device); // FIXME: Remove this when onDeviceIsReady() is called
```
Comment (boundlesscalm) [inline src/app/DeviceSelector.tsx:68]: Confused about the FIXMEs here
Is this still to be done in the future or?

File: src/app/DeviceSelector.tsx [hunk 3]
```diff
@@ -91,12 +95,28 @@ const getBoardControllerVersion = async (
     dispatch(setBoardControllerFirmwareVersion(bcVersion.bc_fw_ver));
 };
 
+const getCurrentBoardControllerConfig = async (
+    dispatch: AppDispatch,
+    device: Device
+) => {
+    if (!device) {
+        return;
+    }
```
Comment (boundlesscalm) [inline src/app/DeviceSelector.tsx:104]: How can the device here be undefined?

File: src/app/DeviceSelector.tsx (hunk 2 above)
Comment (boundlesscalm) [inline src/app/DeviceSelector.tsx:68]: Guess this isn't important, I just don't understand the reason for why not.
But you can also wrap this within dispatch and change the definition for the functions to be
const x = (y) => dispatch => {}

**Review by boundlesscalm — COMMENTED** [review]
I also noticed that you use `{}` as the default/empty state for the hardwareConfig.
I don't know the internals but just wanted to know if this was on purpose.
There were some shorthand booleans related to the hardwareConfig but not exactly this type. I'm just wary because `{}` would equate to true.

**Comment by greg-fer** [comment]
Doc: https://nordicsemi.atlassian.net/browse/NCD-722

@cybic , what exactly needs to be documented in the Board Controller docs?

**Comment by cybic** [comment]
> Doc: https://nordicsemi.atlassian.net/browse/NCD-722
> 
> @cybic , what exactly needs to be documented in the Board Controller docs?

The Board Configurator will now read the config from the board and merge it with the defaults from the board definition file.

File: src/app/DeviceSelector.tsx (hunk 2 above)
Comment (cybic) [reply src/app/DeviceSelector.tsx:68]: I'm not quite sure, but I think so.

**Comment by cybic** [comment]
> I also noticed that you use `{}` as the default/empty state for the hardwareConfig. I don't know the internals but just wanted to know if this was on purpose. There were some shorthand booleans related to the hardwareConfig but not exactly this type. I'm just wary because `{}` would equate to true.

It was done on purpose, and I don't think it's a problem it being truthy. Anyway I've altered the default return to a more correct object (The pins map should always be present, I think)

File: src/app/DeviceSelector.tsx (hunk 3 above)
Comment (cybic) [reply src/app/DeviceSelector.tsx:101]: Removed the check - I came to the same conclusion as you.

File: src/app/DeviceSelector.tsx (hunk 2 above)
Comment (cybic) [reply src/app/DeviceSelector.tsx:68]: I have looked into this briefly earlier with Ådne, although I didn't change it. I'll look into it later on.

**Pull request merged**